README.md
# ai_progress

![totem](https://raw.githubusercontent.com/pdliuw/pdliuw.github.io/master/images/totem_four_logo.jpg)

-----

|[English Document](https://github.com/pdliuw/ai_progress/blob/master/README_EN.md)|[中文文档](https://github.com/pdliuw/ai_progress)|
|:-|:-|

##### ai_progress 支持:

ai_progress: [![pub package](https://img.shields.io/pub/v/ai_progress.svg)](https://pub.dev/packages/ai_progress)

* 1. 圆环进度
* 2. 扇形进度
* 3. 线性进度
* 4. 步进进度
* 5. 仪表盘进度

## Effect

|iOS-ai_progress|Android-ai_progress|
|:-|:-|
|![ios](https://github.com/pdliuw/ai_progress/blob/master/example/gif/ai_progress_ios.gif)|![android](https://github.com/pdliuw/ai_progress/blob/master/example/gif/ai_progress_android.gif)|
|:-|:-|

|macOS-ai_progress|Web-ai_progress|
|:-|:-|
|![macOS](https://github.com/pdliuw/ai_progress/blob/master/example/gif/ai_progress_macOS.gif)|![web](https://github.com/pdliuw/ai_progress/blob/master/example/gif/ai_progress_web.gif)|
|:-|:-|



## 1.安装

使用当前包作为依赖库

### 1. 依赖此库

在文件 'pubspec.yaml' 中添加

[![pub package](https://img.shields.io/pub/v/ai_progress.svg)](https://pub.dev/packages/ai_progress)

```

dependencies:

  ai_progress: ^version

```

或者以下方式依赖

```
dependencies:

  # ai_progress package.
  ai_progress:
    git:
      url: https://github.com/pdliuw/ai_progress.git

```

### 2. 安装此库

你可以通过下面的命令行来安装此库

```

$ flutter pub get


```

你也可以通过项目开发工具通过可视化操作来执行上述步骤

### 3. 导入此库

现在，在你的Dart编辑代码中，你可以使用：

```

import 'package:ai_progress/ai_progress.dart';

```

## 2.使用

### 1. 圆环进度-方形帽

|ai_progress: circular-square|
|:-|
|![circular-square](https://github.com/pdliuw/ai_progress/blob/master/example/gif/ai_progress_circular-square.gif)|
|:-|

```
    
            Stack(
              alignment: Alignment.center,
              children: <Widget>[
                Container(
                  width: 150,
                  height: 150,
                  padding: EdgeInsets.all(5),
                  child: CircularProgressIndicator(
                    value: _segmentValue / 10,
                    strokeWidth: 10.0,
                    valueColor: _colorTween,
                  ),
                ),
                Text("${_segmentValue / 10 * 100}%"),
              ],
            ),
    
```


### 2. 圆环进度-圆形帽

|ai_progress: circular-round|
|:-|
|![circular-round](https://github.com/pdliuw/ai_progress/blob/master/example/gif/ai_progress_circular-round.gif)|
|:-|

```
    
                Stack(
                  alignment: Alignment.center,
                  children: <Widget>[
                    Container(
                      width: 150,
                      height: 150,
                      padding: EdgeInsets.all(5),
                      child: AirCircularStateProgressIndicator(
                        size: Size(150, 150),
                        value: _segmentValue / 10 * 100, //1~100
                        pathColor: Colors.white,
                        valueColor:
                            ColorTween(begin: Colors.grey, end: Colors.blue)
                                .transform(_segmentValue / 10),
                        pathStrokeWidth: 10.0,
                        valueStrokeWidth: 10.0,
                        useCenter: false,
                        filled: false,
                      ),
                    ),
                    Text("${_segmentValue / 10 * 100}%"),
                  ],
                ),
    
```

### 3. 扇形进度

|ai_progress: circular-arc|
|:-|
|![circular-arc](https://github.com/pdliuw/ai_progress/blob/master/example/gif/ai_progress_circular-arc.gif)|
|:-|


```
    
                Stack(
                  alignment: Alignment.center,
                  children: <Widget>[
                    Container(
                      width: 150,
                      height: 150,
                      padding: EdgeInsets.all(5),
                      child: AirCircularStateProgressIndicator(
                        size: Size(150, 150),
                        value: _segmentValue / 10 * 100, //1~100
                        pathColor: Colors.white,
                        valueColor:
                            ColorTween(begin: Colors.grey, end: Colors.blue)
                                .transform(_segmentValue / 10),
                        pathStrokeWidth: 10.0,
                        valueStrokeWidth: 10.0,
                        useCenter: true,
                        filled: true,
                      ),
                    ),
                    Text("${_segmentValue / 10 * 100}%"),
                  ],
                ),
    
```


### 4. 线性进度-方形帽

|ai_progress: linear-square|
|:-|
|![linear-square](https://github.com/pdliuw/ai_progress/blob/master/example/gif/ai_progress_linear-square.gif)|
|:-|

```
    
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    Container(
                      width: 150,
                      padding: EdgeInsets.all(5),
                      child: LinearProgressIndicator(
                        value: _segmentValue / 10,
                        valueColor: _colorTween,
                      ),
                    ),
                    Text("${_segmentValue / 10 * 100}%"),
                  ],
                ),
    
```

### 5. 线性进度-圆形帽

|ai_progress: linear-round|
|:-|
|![linear-round](https://github.com/pdliuw/ai_progress/blob/master/example/gif/ai_progress_linear-round.gif)|
|:-|


```
    
                    Stack(
                      alignment: Alignment.center,
                      children: <Widget>[
                        Container(
                          width: 150,
                          height: 50,
                          padding: EdgeInsets.all(5),
                          child: AirLinearStateProgressIndicator(
                            size: Size(150, 150),
                            value: _segmentValue / 10 * 100, //1~100
                            valueColor:
                                ColorTween(begin: Colors.grey, end: Colors.blue)
                                    .transform(_segmentValue / 10),
                            pathStrokeWidth: 10.0,
                            valueStrokeWidth: 10.0,
                            roundCap: true,
                          ),
                        ),
                        Text("${_segmentValue / 10 * 100}%"),
                      ],
                    ),
    
```

### 6. 步进进度-normal

|ai_progress: step-normal|
|:-|
|![step-normal](https://github.com/pdliuw/ai_progress/blob/master/example/gif/ai_progress_step-normal.gif)|
|:-|

```

    
                    Row(
                      children: [
                        Container(
                          width: 90,
                          height: 50,
                          padding: EdgeInsets.all(0),
                          child: AirStepStateProgressIndicator(
                            size: Size(150, 150),
                            stepCount: _segmentChildren.length,
                            stepValue: _segmentValue,
                            valueColor:
                                ColorTween(begin: Colors.grey, end: Colors.blue)
                                    .transform(_segmentValue / 10),
                            pathStrokeWidth: 10.0,
                            valueStrokeWidth: 10.0,
                          ),
                        ),
                        Text("${_segmentValue / 10 * 100}%"),
                      ],
                    ),
    
```

### 7. 步进进度-larger

|ai_progress: step-larger|
|:-|
|![step-larger](https://github.com/pdliuw/ai_progress/blob/master/example/gif/ai_progress_step-larger.gif)|
|:-|

```
   
                   Row(
                     children: [
                       Container(
                         width: 250,
                         height: 50,
                         padding: EdgeInsets.all(0),
                         child: AirStepStateProgressIndicator(
                           size: Size(150, 150),
                           stepCount: _segmentChildren.length,
                           stepValue: _segmentValue,
                           valueColor:
                               ColorTween(begin: Colors.grey, end: Colors.blue)
                                   .transform(_segmentValue / 10),
                           pathStrokeWidth: 30.0,
                           valueStrokeWidth: 30.0,
                         ),
                       ),
                       Text("${_segmentValue / 10 * 100}%"),
                     ],
                   ), 
    
```

### 8. 步进进度-圆形帽

|ai_progress: step-round|
|:-|
|![step-round](https://github.com/pdliuw/ai_progress/blob/master/example/gif/ai_progress_step-round.gif)|
|:-|

```
   
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    Container(
                      clipBehavior: Clip.antiAlias,
                      decoration: ShapeDecoration(
                        shape: RoundedRectangleBorder(
                            borderRadius: BorderRadius.only(
                          topLeft: Radius.circular(45),
                          bottomLeft: Radius.circular(45),
                          topRight: Radius.circular(45),
                          bottomRight: Radius.circular(45),
                        )),
                      ),
                      width: 220,
                      height: 30.0,
                      child: AirStepStateProgressIndicator(
                        size: Size(150, 220),
                        stepCount: _segmentChildren.length,
                        stepValue: _segmentValue,
                        valueColor:
                            ColorTween(begin: Colors.grey, end: Colors.blue)
                                .transform(_segmentValue / 10),
                        pathStrokeWidth: 30.0,
                        valueStrokeWidth: 30.0,
                      ),
                    ),
                    Text("${_segmentValue / 10 * 100}%"),
                  ],
                ),
    
```


### 9. 仪表盘进度-方形帽

|ai_progress: dashboard-square|
|:-|
|![dashboard-square](https://github.com/pdliuw/ai_progress/blob/master/example/gif/ai_progress_dashboard-square.gif)|
|:-|

```
   
                Stack(
                  alignment: Alignment.center,
                  children: <Widget>[
                    Container(
                      width: 150,
                      height: 150,
                      padding: EdgeInsets.all(5),
                      child: AirDashboardStateProgressIndicator(
                        size: Size(150, 150),
                        value: _segmentValue / 10 * 100, //1~100
                        valueColor:
                            ColorTween(begin: Colors.grey, end: Colors.blue)
                                .transform(_segmentValue / 10),
                        pathStrokeWidth: 10,
                        valueStrokeWidth: 10,
                        gapDegree: 60,
                        roundCap: false,
                      ),
                    ),
                    Text("${_segmentValue / 10 * 100}%"),
                  ],
                ),
    
```

### 10. 仪表盘进度-圆形帽

|ai_progress: dashboard-round|
|:-|
|![dashboard-round](https://github.com/pdliuw/ai_progress/blob/master/example/gif/ai_progress_dashboard-round.gif)|
|:-|

```
   
                Stack(
                  alignment: Alignment.center,
                  children: <Widget>[
                    Container(
                      width: 150,
                      height: 150,
                      padding: EdgeInsets.all(5),
                      child: AirDashboardStateProgressIndicator(
                        size: Size(150, 150),
                        value: _segmentValue / 10 * 100, //1~100
                        valueColor:
                            ColorTween(begin: Colors.grey, end: Colors.blue)
                                .transform(_segmentValue / 10),
                        pathStrokeWidth: 10,
                        valueStrokeWidth: 10,
                        gapDegree: 60,
                        roundCap: true,
                      ),
                    ),
                    Text("${_segmentValue / 10 * 100}%"),
                  ],
                ),
    
```

意犹未尽？[点击，查看项目示例](https://github.com/pdliuw/flutter_app_sample)


## LICENSE

    BSD 3-Clause License
    
    Copyright (c) 2020, pdliuw
    All rights reserved.




CHANGELOG.md
## [2.0.0]

* null safety.

## [1.0.0] - Stable

* adapter: flutter version = v1.22
* upgrade: iOS development target from 8.0 to 9.0 

## [0.2.1] - Stable

* Update: example document and resource 

## [0.2.0-pre] - pre

* Add: 'clipBehavior: Clip.antiAlias,' compile 'v0.2.0'

## [0.2.0] - Stable

* Stable

## [0.1.9] - Add

* Add: Step progress round cap style

## [0.1.3] - Update

* Update: document resource

## [0.1.2] - Dashboard

* Add: Dashboard progress

## [0.1.1] - Document

* config document

## [0.1.0] - Stable

* stable version

## [0.0.2] - Document

* add document

## [0.0.1] - initial

* initial release.


test/aiprogress_test.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:ai_progress/ai_progress.dart';

void main() {
  test('adds one to input values', () {
    AirCircularStateProgressIndicator(
      size: Size(20, 20),
    );
  });
}


lib/src/step_state_progress_indicator.dart
import 'package:ai_progress/src/progress_mixin.dart';
import 'package:flutter/material.dart';

/// AirStepStateProgressIndicator
// ignore: must_be_immutable
class AirStepStateProgressIndicator extends StatefulWidget {
  /// Limited
  static const int LIMITED_STEP_MIN_VALUE = 0;

  /// Default values
  static const int DEFAULT_STEP_COUNT = 10;
  static const int DEFAULT_STEP_VALUE = 1;
  static const double DEFAULT_STROKE_WIDTH = 5.0;
  static const double DEFAULT_SPACE_WIDTH = 2.0;

  final Size size;
  final int stepCount;
  final int stepValue;

  double? _pathStrokeWidth;
  double? _valueStrokeWidth;
  double? _spaceWidth;

  Color? _pathColor;
  Color? _valueColor;

  /// constructor
  AirStepStateProgressIndicator({
    required this.size,
    this.stepCount = DEFAULT_STEP_COUNT,
    this.stepValue = DEFAULT_STEP_VALUE,
    double pathStrokeWidth = DEFAULT_STROKE_WIDTH,
    double valueStrokeWidth = DEFAULT_STROKE_WIDTH,
    double spaceWidth = DEFAULT_SPACE_WIDTH,
    Color pathColor = Colors.grey,
    Color valueColor = Colors.green,
  }) {
    assert(stepCount > LIMITED_STEP_MIN_VALUE);
    assert(stepValue >= LIMITED_STEP_MIN_VALUE);
    assert(stepValue <= stepCount);

    //stroke width
    _pathStrokeWidth = pathStrokeWidth;
    _valueStrokeWidth = valueStrokeWidth;

    //space
    _spaceWidth = spaceWidth;

    //color
    _pathColor = pathColor;
    _valueColor = valueColor;
  }
  @override
  State<StatefulWidget> createState() {
    return _StepStateProgressIndicator();
  }
}

/// _StepStateProgressIndicator
class _StepStateProgressIndicator extends State<AirStepStateProgressIndicator> {
  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      size: widget.size,
      painter: StepProgressPainter._(
        shouldRepaint: false,
        stepCount: widget.stepCount,
        stepValue: widget.stepValue,
        pathStrokeWidth: widget._pathStrokeWidth,
        valueStrokeWidth: widget._valueStrokeWidth,
        spaceWidth: widget._spaceWidth,
        pathColor: widget._pathColor,
        valueColor: widget._valueColor,
        roundCap: false,
      ),
    );
  }
}

/// StepProgressPainter
class StepProgressPainter extends CustomPainter with ProgressMixin {
  bool? _shouldRepaint;
  int? _stepCount;
  int? _stepValue;

  double? _pathStrokeWidth;
  double? _valueStrokeWidth;

  double? _spaceWidth;

  Color? _valueColor;
  Color? _pathColor;

  bool? _roundCap;

  /// constructor
  StepProgressPainter._({
    required bool shouldRepaint,
    required int stepCount,
    required int stepValue,
    double? pathStrokeWidth,
    double? valueStrokeWidth,
    double? spaceWidth,
    Color? pathColor,
    Color? valueColor,
    bool? roundCap,
  }) {
    //should repaint
    _shouldRepaint = shouldRepaint;

    //step count
    _stepCount = stepCount;

    //step value
    _stepValue = stepValue;

    //stroke
    _pathStrokeWidth = pathStrokeWidth;
    _valueStrokeWidth = valueStrokeWidth;
    _roundCap = roundCap;

    //
    _spaceWidth = spaceWidth;

    //color
    _pathColor = pathColor;
    _valueColor = valueColor;
  }
  @override
  void paint(Canvas canvas, Size size) {
    //draw progress path
    drawProgressPath(canvas: canvas, paint: _getDrawPathPaint(), size: size);

    //draw progress value
    drawProgressValue(canvas: canvas, paint: _getDrawValuePaint(), size: size);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return _shouldRepaint!;
  }

  @override
  drawProgressPath(
      {required Canvas canvas, required Paint paint, required Size size}) {
    //draw step progress path
    _drawStep(canvas: canvas, paint: paint, size: size, loopCount: stepCount);
  }

  @override
  drawProgressValue(
      {required Canvas canvas, required Paint paint, required Size size}) {
    //draw step progress value
    _drawStep(canvas: canvas, paint: paint, size: size, loopCount: stepValue);
  }

  _drawStep(
      {required Canvas canvas,
      required Paint paint,
      required Size size,
      required int loopCount}) {
    double y = size.height / 2;

    double startY = y;
    double endY = y;
    for (int i = 0; i < loopCount; i++) {
      double startX =
          i * (_getSingleStepWidth(size: size) + _getSingleSpaceWidth());
      double endX = startX + _getSingleStepWidth(size: size);

      Offset startPoint = Offset(startX, startY);
      Offset endPoint = Offset(endX, endY);
      //draw
      canvas.drawLine(startPoint, endPoint, paint);
    }
  }

  Paint _getDrawPathPaint() {
    return Paint()
      ..color = pathColor
      ..style = PaintingStyle.stroke
      ..strokeCap = _roundCap! ? StrokeCap.butt : StrokeCap.butt
      ..strokeJoin = _roundCap! ? StrokeJoin.miter : StrokeJoin.miter
      ..strokeWidth = pathStrokeWidth;
  }

  Paint _getDrawValuePaint() {
    return Paint()
      ..color = valueColor
      ..style = PaintingStyle.stroke
      ..strokeCap = _roundCap! ? StrokeCap.butt : StrokeCap.butt
      ..strokeJoin = _roundCap! ? StrokeJoin.miter : StrokeJoin.miter
      ..strokeWidth = valueStrokeWidth;
  }

  int get stepCount => _stepCount!;

  int get stepValue => _stepValue!;

  Color get valueColor => _valueColor!;
  Color get pathColor => _pathColor!;

  double get pathStrokeWidth => _pathStrokeWidth!;
  double get valueStrokeWidth => _valueStrokeWidth!;

  double _getSingleStepWidth({required Size size}) {
    return (size.width - ((stepCount - 1) * _getSingleSpaceWidth())) /
        stepCount;
  }

  double _getSingleSpaceWidth() {
    return _spaceWidth!;
  }
}


example/lib/main.dart
import 'dart:ui';

import 'package:ai_progress/ai_progress.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        Expanded(
          flex: 1,
          child: MyHomePage(title: 'Flutter Demo Home Page'),
        ),
      ],
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage>
    with SingleTickerProviderStateMixin {
  int _counter = 0;

  static const double MAX = 10.0;
  static const double MIN = 1.0;
  static int _divisions = 99;

  double _progressValue = 1;
  double _slideValue = MAX;

  AnimationController _controller;
  Animation<Color> _colorTween;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(vsync: this);
    _colorTween = _controller.drive(ColorTween(
      begin: Colors.grey,
      end: Colors.green,
    ));
    _controller.value = _progressValue;
  }

  @override
  void dispose() {
    super.dispose();
    _controller.dispose();
    Paint paint = Paint();

    paint..isAntiAlias = true;
  }

  Map<int, Widget> _segmentChildren = {
    1: Text("1"),
    2: Text("2"),
    3: Text("3"),
    4: Text("4"),
    5: Text("5"),
    6: Text("6"),
    7: Text("7"),
    8: Text("8"),
    9: Text("9"),
    10: Text("10"),
  };
  int _segmentValue = 1;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: SingleChildScrollView(
        child: Column(
          children: [
            Row(
              children: [
                Expanded(
                  child: CupertinoSegmentedControl(
                    padding: EdgeInsets.all(5),
                    children: _segmentChildren,
                    onValueChanged: (int index) {
                      setState(() {
                        _segmentValue = index;
                        _controller.value = index * 10 / 100;
                      });
                    },
                    groupValue: _segmentValue,
                  ),
                ),
              ],
            ),
            Align(
              child: Text("$_segmentValue"),
            ),
            Slider(
              min: MIN,
              max: MAX,
              value: _segmentValue.toDouble(),
              onChanged: (double newValue) {
                setState(() {
                  _segmentValue = newValue.toInt();
                  _controller.value = newValue;
                });
              },
            ),
            Row(
              children: [
                Spacer(),
                Stack(
                  alignment: Alignment.center,
                  children: <Widget>[
                    Container(
                      width: 150,
                      height: 150,
                      padding: EdgeInsets.all(5),
                      child: CircularProgressIndicator(
                        value: _segmentValue / 10,
                        strokeWidth: 10.0,
                        valueColor: _colorTween,
                      ),
                    ),
                    Text("${_segmentValue / 10 * 100}%"),
                  ],
                ),
                Spacer(),
                Stack(
                  alignment: Alignment.center,
                  children: <Widget>[
                    Container(
                      width: 150,
                      height: 150,
                      padding: EdgeInsets.all(5),
                      child: AirDashboardStateProgressIndicator(
                        size: Size(150, 150),
                        value: _segmentValue / 10 * 100, //1~100
                        valueColor:
                            ColorTween(begin: Colors.grey, end: Colors.blue)
                                .transform(_segmentValue / 10),
                        pathStrokeWidth: 10,
                        valueStrokeWidth: 10,
                        gapDegree: 60,
                        roundCap: true,
                      ),
                    ),
                    Text("${_segmentValue / 10 * 100}%"),
                  ],
                ),
                Spacer(),
              ],
            ),
            //圆环、扇形样式的进度
            Row(
              children: [
                Spacer(),
                Stack(
                  alignment: Alignment.center,
                  children: <Widget>[
                    Container(
                      width: 150,
                      height: 150,
                      padding: EdgeInsets.all(5),
                      child: AirCircularStateProgressIndicator(
                        size: Size(150, 150),
                        value: _segmentValue / 10 * 100, //1~100
                        pathColor: Colors.white,
                        valueColor:
                            ColorTween(begin: Colors.grey, end: Colors.blue)
                                .transform(_segmentValue / 10),
                        pathStrokeWidth: 10.0,
                        valueStrokeWidth: 10.0,
                        useCenter: false,
                        filled: false,
                        roundCap: true,
                      ),
                    ),
                    Text("${_segmentValue / 10 * 100}%"),
                  ],
                ),
                Spacer(),
                Stack(
                  alignment: Alignment.center,
                  children: <Widget>[
                    Container(
                      width: 150,
                      height: 150,
                      padding: EdgeInsets.all(5),
                      child: AirCircularStateProgressIndicator(
                        size: Size(150, 150),
                        value: _segmentValue / 10 * 100, //1~100
                        pathColor: Colors.white,
                        valueColor:
                            ColorTween(begin: Colors.grey, end: Colors.blue)
                                .transform(_segmentValue / 10),
                        pathStrokeWidth: 10.0,
                        valueStrokeWidth: 10.0,
                        useCenter: true,
                        filled: true,
                        roundCap: true,
                      ),
                    ),
                    Text("${_segmentValue / 10 * 100}%"),
                  ],
                ),
                Spacer(),
              ],
            ),
            //无进度值的加载
            CircularProgressIndicator(),
            //线性、步进样式的进度
            Row(
              children: [
                Spacer(),
                Stack(
                  alignment: Alignment.center,
                  children: <Widget>[
                    Container(
                      width: 150,
                      height: 50,
                      padding: EdgeInsets.all(5),
                      child: AirLinearStateProgressIndicator(
                        size: Size(150, 150),
                        value: _segmentValue / 10 * 100, //1~100
                        valueColor:
                            ColorTween(begin: Colors.grey, end: Colors.blue)
                                .transform(_segmentValue / 10),
                        pathStrokeWidth: 10.0,
                        valueStrokeWidth: 10.0,
                        roundCap: true,
                      ),
                    ),
                    Text("${_segmentValue / 10 * 100}%"),
                  ],
                ),
                Spacer(),
                Row(
                  children: [
                    Container(
                      width: 90,
                      height: 10,
                      padding: EdgeInsets.all(0),
                      child: AirStepStateProgressIndicator(
                        size: Size(150, 150),
                        stepCount: _segmentChildren.length,
                        stepValue: _segmentValue,
                        valueColor:
                            ColorTween(begin: Colors.grey, end: Colors.blue)
                                .transform(_segmentValue / 10),
                        pathStrokeWidth: 10.0,
                        valueStrokeWidth: 10.0,
                      ),
                    ),
                    Text("${_segmentValue / 10 * 100}%"),
                  ],
                ),
                Spacer(),
              ],
            ),
            //Step progress round and square cap
            Column(
              crossAxisAlignment: CrossAxisAlignment.center,
              children: [
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    Container(
                      width: 220,
                      height: 30,
                      padding: EdgeInsets.all(0),
                      child: AirStepStateProgressIndicator(
                        size: Size(150, 150),
                        stepCount: _segmentChildren.length,
                        stepValue: _segmentValue,
                        valueColor:
                            ColorTween(begin: Colors.grey, end: Colors.blue)
                                .transform(_segmentValue / 10),
                        pathStrokeWidth: 30.0,
                        valueStrokeWidth: 30.0,
                      ),
                    ),
                    Text("${_segmentValue / 10 * 100}%"),
                  ],
                ),
                Divider(),
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    Container(
                      clipBehavior: Clip.antiAlias,
                      decoration: ShapeDecoration(
                        shape: RoundedRectangleBorder(
                            borderRadius: BorderRadius.only(
                          topLeft: Radius.circular(45),
                          bottomLeft: Radius.circular(45),
                          topRight: Radius.circular(45),
                          bottomRight: Radius.circular(45),
                        )),
                      ),
                      width: 220,
                      height: 30.0,
                      child: AirStepStateProgressIndicator(
                        size: Size(150, 220),
                        stepCount: _segmentChildren.length,
                        stepValue: _segmentValue,
                        valueColor:
                            ColorTween(begin: Colors.grey, end: Colors.blue)
                                .transform(_segmentValue / 10),
                        pathStrokeWidth: 30.0,
                        valueStrokeWidth: 30.0,
                      ),
                    ),
                    Text("${_segmentValue / 10 * 100}%"),
                  ],
                ),
              ],
            ),
            Divider(),
            //linear square progress
            Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                Container(
                  width: 150,
                  padding: EdgeInsets.all(5),
                  child: LinearProgressIndicator(
                    value: _segmentValue / 10,
                    valueColor: _colorTween,
                  ),
                ),
                Text("${_segmentValue / 10 * 100}%"),
              ],
            ),
          ],
        ),
      ),
    );
  }
}
